### PR TITLE
nixos/atuin: fix `database.createLocally`, harden systemd unit

### PR DESCRIPTION
**Pull request by h7x4**
###### Description of changes

The earlier module would start up a local postgres server, even if you specified `database.createLocally = false`. This PR cleans up this functionality a bit, doesn't implicitly create a new postgres server, and uses the socket as default. In addition, the systemd unit has been hardened.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

I've run `nix-build -A nixosTests.atuin` as well.

### CODE REVIEW

File: nixos/modules/services/misc/atuin.nix [hunk 1]
```diff
@@ -50,16 +48,28 @@ in
...
     };
   };
 
   config = mkIf cfg.enable {
-
-    # enable postgres to host atuin db
-    services.postgresql = {
+    assertions = [
+      {
+        assertion = !config.services.postgresql.enable -> !cfg.database.createLocally;
```
[the remark this answers is not in the document]
Comment (h7x4) [reply nixos/modules/services/misc/atuin.nix:67]: Good observation! I'll change it right away!

File: nixos/modules/services/misc/atuin.nix [hunk 2]
```diff
@@ -73,28 +83,63 @@ in
...
       };
 
       environment = {
         ATUIN_HOST = cfg.host;
         ATUIN_PORT = toString cfg.port;
         ATUIN_MAX_HISTORY_LENGTH = toString cfg.maxHistoryLength;
-        ATUIN_OPEN_REGISTRATION = boolToString cfg.openRegistration;
-        ATUIN_DB_URI =  mkIf cfg.database.createLocally "postgresql:///atuin";
+        ATUIN_OPEN_REGISTRATION = lib.boolToString cfg.openRegistration;
+        ATUIN_DB_URI = cfg.database.uri;
```
[the remark this answers is not in the document]
Comment (h7x4) [reply nixos/modules/services/misc/atuin.nix:137]: The `mkIf` causes the database uri not to be set at all unless you choose to create it locally. You would have to set all you settings in `services.atuin`, except for the database uri, which you would have to set in `systemd.services.atuin.environment.ATUIN_DB_URI`.

I've created a new option `services.atuin.database.uri`. If you set `database.createLocally = false`, you're probably going to set `database.uri` as well.

[the remark this answers is not in the document]
Comment (h7x4) [reply nixos/modules/services/misc/atuin.nix:137]: I didn't know atuin had an SQLite backend. [According to their docs, a postgres database is required](https://atuin.sh/docs/self-hosting)

[the remark this answers is not in the document]
Comment (h7x4) [reply nixos/modules/services/misc/atuin.nix:137]: I had a go at it just to make sure I hadn't missed any hidden features, and it gave me this:

```nix
services.atuin = {
  enable = true;
  database.createLocally = false;
};
```

```
Error: could not load server settings
Caused by:
    failed to deserialize: missing field `db_uri`
Location:
    atuin-server/src/settings.rs:69:26
```

If there is any SQLite support, it's at least not enabled by letting the setting be empty.

[the remark this answers is not in the document]
Comment (h7x4) [reply nixos/modules/services/misc/atuin.nix:137]: Yep! No worries. It's nice to clear up any confusion, could've been the case that SQLite actually were supported and I'd accidentally removed support here :smile:

[the remark this answers is not in the document]
Comment (h7x4) [reply nixos/modules/services/misc/atuin.nix:137]: I'll mark it as resolved

**Review by amarshall — APPROVED** [review]
Code looks reasonable to me. I have not done any testing as I don’t use this service.

**Pull request merged**